### ISSUE TITLE
Bug: Export from eslint-plugin does not match types expected by eslint

### ISSUE THREAD

**Issue by skondrashov** [typescript-eslint/typescript-eslint#11038]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

The export from `@typescript-eslint/eslint-plugin` has multiple types that seem to be incompatible with the type of the parameter of `fixupPluginRules` exported from `@eslint/compat`. This is the same type that's expected by the config in the `plugins` field, so even if `fixupPluginRules` isn't used (which it appears it no longer has to, but my project did), some of these issues still apply.

### 1

This first one only appears with `strictNullChecks: true` (or `strict: true`) in tsconfig, as a result of a mismatch between:

```typescript
// eslint/lib/types/index.d.ts
interface BaseConfig {
  parser?: string | undefined;
}
```

```typescript
// packages/utils/src/ts-eslint/Config.ts
interface BaseConfig {
  parser?: string | null;
}
```

### 2

This one is convoluted, but ultimately down to a mismatch between:

```typescript
// (this is in eslint/rewrite) rewrite/packages/core/src/types.ts
interface RulesMetaDocs {
  recommended?: boolean | undefined;
}
```

```typescript
// @typescript-eslint/eslint-plugin/rules.d.ts
interface ESLintPluginDocs {
  recommended?: RuleRecommendation | RuleRecommendationAcrossConfigs<unknown[]>;
}
```

There's at least one more, though I get the feeling there's probably a bunch more, but it takes a while to find the actual source of the error so I'll leave at those two for now.

I know `rules.d.ts` says:
```
We purposely don't generate types for our plugin because TL;DR:
1) there's no real reason that anyone should do a typed import of our rules,
2) it would require us to change our code so there aren't as many inferred types
```

But real reason or not (my reason is I'm generating a config in a library that is exported to multiple apps and I didn't see why it shouldn't be in typescript), the types are wrong. But I could see this being a low priority, and/or needing to be opened with the eslint folks instead. Just thought I'd report since I didn't find any issues about this.

### Reproduction Repository Link

https://github.com/skondrashov/eslint-bug-repro

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn lint`

Really the minimal example is:
```typescript
import { fixupPluginRules } from '@eslint/compat'
import typescriptEslint from '@typescript-eslint/eslint-plugin'

fixupPluginRules(typescriptEslint)
```

but the repo I provided more closely resembles what I'm actually trying to do, which is export parts of the eslint config (from a library) into eslint configs in different apps.

### Versions

| package                            | version   |
| ---------------------------------- | --------- |
| `@eslint/compat`                   | `1.2.8`   |
| `@eslint/eslintrc`                 | `3.3.1`   |
| `@eslint/js`                       | `9.24.0`  |
| `@types/node`                      | `22.14.0` |
| `@typescript-eslint/eslint-plugin` | `8.29.1`  |
| `@typescript-eslint/parser`        | `8.29.1`  |
| `eslint`                           | `9.24.0`  |
| `typescript`                       | `5.8.3`   |

**Comment by skondrashov** [comment]
It works, thank you!
